### PR TITLE
storage: fix overaggressive raft log truncation

### LINKED ISSUE #8629: stability: repeatedly applying snapshot for the same range
Running `block_writer` against a local 4 node cluster, I see a snapshot being applied repeatedly for 1 range on one node:

```
~/Development/go/src/github.com/cockroachdb/cockroach pmattis/raft-logging-redux grep 'applied Raft snapshot for range 7 ' cockroach-data/3/logs/0.stderr | wc -l
      73
```

The only replica changes for this range are:

```
~/Development/go/src/github.com/cockroachdb/cockroach pmattis/raft-logging-redux grep 'range=7' cockroach-data/1/logs/0.stderr | egrep 'ADD|REMOVE'
I160817 19:53:52.667525 storage/replica.go:1500  store=1:1 range=7 [/Table/51-/Table/51/1/2319145621337716466/"4c3241bc-22e1-4c42-8996-f40f3f3ef665"/10028): proposing ADD_REPLICA {NodeID:4 StoreID:4 ReplicaID:4} for range 7: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:3 StoreID:2 ReplicaID:2} {NodeID:2 StoreID:3 ReplicaID:3} {NodeID:4 StoreID:4 ReplicaID:4}]
I160817 19:53:52.772124 storage/replica.go:1500  store=1:1 range=7 [/Table/51-/Table/51/1/2319145621337716466/"4c3241bc-22e1-4c42-8996-f40f3f3ef665"/10028): proposing REMOVE_REPLICA {NodeID:3 StoreID:2 ReplicaID:2} for range 7: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:4 StoreID:4 ReplicaID:4} {NodeID:2 StoreID:3 ReplicaID:3}]
I160817 19:54:49.981856 storage/replica.go:1500  store=1:1 range=7 [/Table/51-/Table/51/1/2319145621337716466/"4c3241bc-22e1-4c42-8996-f40f3f3ef665"/10028): proposing ADD_REPLICA {NodeID:3 StoreID:2 ReplicaID:5} for range 7: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:4 StoreID:4 ReplicaID:4} {NodeID:2 StoreID:3 ReplicaID:3} {NodeID:3 StoreID:2 ReplicaID:5}]
I160817 19:54:51.671439 storage/replica.go:1500  store=1:1 range=7 [/Table/51-/Table/51/1/2319145621337716466/"4c3241bc-22e1-4c42-8996-f40f3f3ef665"/10028): proposing REMOVE_REPLICA {NodeID:2 StoreID:3 ReplicaID:3} for range 7: [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:4 StoreID:4 ReplicaID:4} {NodeID:3 StoreID:2 ReplicaID:5}]
```

### PR DESCRIPTION
**Pull request by petermattis**
Fixes #8629.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/8636)

### CODE REVIEW

File: storage/raft_log_queue.go [hunk 1]
```diff
@@ -114,28 +114,31 @@ func getTruncatableIndexes(r *Replica) (uint64, uint64, error) {
...
+// snapshots. See #8629.
 func computeTruncatableIndex(
 	raftStatus *raft.Status,
 	raftLogSize, targetSize int64,
 	firstIndex uint64,
 ) uint64 {
 	truncatableIndex := raftStatus.Commit
 	behindIndexes := getBehindIndexes(raftStatus)
-	if raftLogSize <= targetSize {
+	if raftLogSize <= targetSize && len(behindIndexes) > 0 {
```
Comment (spencerkimball) [inline storage/raft_log_queue.go]: By the way, why in the heck are we computing a unique array of behind indexes in `getBehindIndexes`? That's just silly. Could you remove that extra complication?

**Comment by spencerkimball** [comment]
LGTM

File: storage/raft_log_queue.go (hunk 1 above)
Comment (petermattis) [inline storage/raft_log_queue.go]: Done.

**Comment by cuongdo** [comment]
:lgtm:

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/raft_log_queue.go, line 135 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8636#-KPSd00PD0oG4CcrT-qy-r1-135:-KPSqwCOnD1rnnsDu-tA:b-9gghuo) ([raw file](https://github.com/cockroachdb/cockroach/blob/14bf1a51450e7e327d4c4747430c96722bee4eb3/storage/raft_log_queue.go#L135)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> By the way, why in the heck are we computing a unique array of behind indexes in `getBehindIndexes`? That's just silly. Could you remove that extra complication?
</details>
+1 the uniqueness isn't needed, but the sorting is

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8636#-:-KPSrDQlt-yeFrZaRSZ5:bnfp4nl)*

**Comment by cuongdo** [comment]
*[storage/raft_log_queue.go, line 156 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8636#-KPSsPCYKgU2NyeMcDM2:-KPSsPCZVjy8HOvOvTGN:bth20zr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e74baa8b71752a457c283bf5548760b24fdb3ae3/storage/raft_log_queue.go#L156)):*
> ```Go
> 	for _, progress := range raftStatus.Progress {
> 		index := progress.Match
> 		if index < raftStatus.Commit && behind > index {
> ```

is `index < raftStatus.Commit` needed? `behind` starts at `raftStatus.Commit` and can only decrease from there because of the `behind > index` condition

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8636)*

**Comment by cuongdo** [comment]
*[storage/raft_log_queue.go, line 156 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8636#-KPSsPCYKgU2NyeMcDM2:-KPSt6AdD_8jG4bFGWLF:b42hu9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/e74baa8b71752a457c283bf5548760b24fdb3ae3/storage/raft_log_queue.go#L156)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> is `index < raftStatus.Commit` needed? `behind` starts at `raftStatus.Commit` and can only decrease from there because of the `behind > index` condition
</details>
also, `behind > index` would read more naturally to me as `index < behind`, but this is a nit, so your call

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8636)*

**Comment by petermattis** [comment]
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[storage/raft_log_queue.go, line 156 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8636#-KPSsPCYKgU2NyeMcDM2:-KPStbHyUmVMLO4rND1w:bvaiovb) ([raw file](https://github.com/cockroachdb/cockroach/blob/e74baa8b71752a457c283bf5548760b24fdb3ae3/storage/raft_log_queue.go#L156)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> also, `behind > index` would read more naturally to me as `index < behind`, but this is a nit, so your call
</details>
What if `progress.Match > raftStatus.Commit`? I'd rather be conservative here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8636)*

**Comment by cuongdo** [comment]
*[storage/raft_log_queue.go, line 156 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8636#-KPSsPCYKgU2NyeMcDM2:-KPStxm0gbxdKFWmX7mu:b-7fsa8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/e74baa8b71752a457c283bf5548760b24fdb3ae3/storage/raft_log_queue.go#L156)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> What if `progress.Match > raftStatus.Commit`? I'd rather be conservative here.
</details>
suppose:

```
progress.Match = 2
raftStatus.Commit = 1
```

`behind` would start at 1. `behind` (1) isn't greater than `progress.Match` (2), so `behind isn't updated. However, being conservative with this code is probably a good thing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8636)*

**Comment by cuongdo** [comment]
still :lgtm:

---

Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8636#-:-KPSurr9TXm4uwNTrpR7:b-5fsfav)*

**Pull request merged**